walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,251 @@
+      splitEvents.add(TimeRange.fromStartEnd(prevEndTime, range.end(), false));
+    }
+    return splitEvents;
+  }
+
+  /** 
+   * Counts the number of optional attendees that can not make it to the meeting at
+   * the time range and compares it to the minimum number of optional attendees that
+   * can not make it for other time ranges.
+   * @param bestTimes collection which has the TimeRanges of meetings that has 
+   * the minOverlap attendees 
+   * @param events events of the day
+   * @param optional optional attendees for the meeting 
+   * @param range range in question to see how many optional attendees can not make it 
+   * @param minOverlaps current minimum number of optional attendees that can not make
+   * it to the meeting at the TimeRange
+   * @return the minimum of current range's overlaps and minOverlaps
+   */
+  private int optionalCounter(Collection<TimeRange> bestTimes,
+		  Collection<Event> events, Collection<String> optional,
+		  TimeRange range, int minOverlaps) {
+    int counterOverlaps = 0;
+    for (String opAttendee : optional) {
+      Collection<Event> eventsWithAttendee = eventFinderWithAttendee(events, opAttendee);
+      for (Event e : eventsWithAttendee){
+        if (e.getWhen().overlaps(range)) {
+          counterOverlaps += 1;
+        }
+      }
+    }
+
+    if (counterOverlaps < minOverlaps) {
+      bestTimes.clear();
+      bestTimes.add(range);
+      return counterOverlaps;
+    }
+    else if (counterOverlaps == minOverlaps) {
+      bestTimes.add(range);
+    }
+    return minOverlaps;
+  }
+
+  /** 
+   * Determines the TimeRanges where the mandatory and as many optional
+   * attendees as possible can attend the meeting.
+   * @param events events of the day 
+   * @param request includes mandatory and optional attendees, and the duration 
+   * of the meeting 
+   * @return collection of TimeRanges where all mandatory and the greatest 
+   * number of optional attendees can be at the meeting
+   */
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest
+		  request) {
+    long duration = request.getDuration();
+ 
+    Collection<String> mandatory = request.getAttendees();
+    Collection<String> optional = request.getOptionalAttendees();
+    
+    // sort the events to be in ascending order by start time
+    List<Event> eventsList = new ArrayList<Event>(events); 
+    Collections.sort(eventsList, Event.ORDER_BY_START); 
+    List<TimeRange> possibleTimes = new ArrayList<>();
+
+    /**
+     * If the duration of meeting is longer than a day or less than 0 minutes,
+     * then there are no meeting times. 
+     */
+    if (duration > TimeRange.getTimeInMinutes(23, 59) || duration < 0) {
+      return possibleTimes;
+    }
+    
+    /* If there are no events for the day, then the meeting can be any time. */
+    if (events.isEmpty()) {
+      possibleTimes.add(TimeRange.WHOLE_DAY);
+      return possibleTimes;
+    }
+    
+    if (!optional.isEmpty()) {
```

Comment:
This `if` is getting too long and a bit hard to follow. I'd move the `option.isEmpty()` case here and return early, to reduce the nested if.